typing_extensions/src_py3/typing_extensions.py

```diff
@@ -985,6 +997,188 @@ def __new__(cls, *args, **kwds):
         Protocol.__doc__ = Protocol.__doc__.format(bases="Protocol, Generic[T]" if
                                                    OLD_GENERICS else "Protocol[T]")
 
+
+if PEP_560:
+    # We need to rewrite these from scratch.
+    del _ProtocolMeta, Protocol
+
+    from typing import _type_check, _GenericAlias, _collect_type_vars
+
+    class _ProtocolMeta(abc.ABCMeta):
+        # This metaclass is a bit unfortunate and exists only because of the lack
+        # of __instancehook__.
+        def __instancecheck__(cls, instance):
+            # We need this method for situations where attributes are
+            # assigned in __init__.
+            if ((not getattr(cls, '_is_protocol', False) or
+                    _callable_members_only(cls)) and
+                    issubclass(instance.__class__, cls)):
+                return True
+            if cls._is_protocol:
+                if all(hasattr(instance, attr) and
+                        (not callable(getattr(cls, attr, None)) or
+                         getattr(instance, attr) is not None)
+                        for attr in _get_protocol_attrs(cls)):
+                    return True
+            return super().__instancecheck__(instance)
+
+
+    class Protocol(metaclass=_ProtocolMeta):
+        # There is quite a few overlapping code with typing.Generic.
+        # Unfortunarely it is hard to avoid this while these live in two different modules.
+        # The duplicated code will be removed when Protocol is moved to typing.
+        """Base class for protocol classes. Protocol classes are defined as::
+
+            class Proto(Protocol):
+                def meth(self) -> int:
+                    ...
+
+        Such classes are primarily used with static type checkers that recognize
+        structural subtyping (static duck-typing), for example::
+
+            class C:
+                def meth(self) -> int:
+                    return 0
+
+            def func(x: Proto) -> int:
+                return x.meth()
+
+            func(C())  # Passes static type check
+
+        See PEP 544 for details. Protocol classes decorated with
+        @typing_extensions.runtime act as simple-minded runtime protocol that checks
+        only the presence of given attributes, ignoring their type signatures.
+
+        Protocol classes can be generic, they are defined as::
```

Comment:
Protocols are not different from normal generics in this aspect, for example:
```python
@runtime_checkable
class GenP(Protocol[T]):
    def meth(self) -> T:
        ...
class C:
    ...
issubclass(C, GenP)  # This is OK, False or True depending on C
issubclass(C, GenP[int])  # This is an error both in mypy and at runtime.
```

Reply:
Still my suggestion stands -- `@runtime` (by whichever name) should complain when the decorated class is a generic protocol. Right? Since the only reason to add `@runtime` is to enable `isinstance()` checks and that is not allowed for generic protocols.

Reply:
Oh wait. After reading [the relevant passage in PEP 544](https://www.python.org/dev/peps/pep-0544/#runtime-decorator-and-narrowing-types-by-isinstance) again I finally realize that `@isinstance(x, GenP)` is still allowed, it's only `@isinstance(x, GenP[int])` that's disallowed. So forget everything I said here. (Except that as I said in a previous comment it's been a long time since I thought about this. :-( )